src/field.rs

```diff
@@ -10,132 +10,279 @@ use arbitrary::Arbitrary;
-    0x4d, 0x9a, 0x29, 0x52, 0xa4, 0x55, 0xaa, 0x49, 0x92, 0x39, 0x72, 0xe4, 0xd5, 0xb7, 0x73, 0xe6,
-    0xd1, 0xbf, 0x63, 0xc6, 0x91, 0x3f, 0x7e, 0xfc, 0xe5, 0xd7, 0xb3, 0x7b, 0xf6, 0xf1, 0xff, 0xe3,
-    0xdb, 0xab, 0x4b, 0x96, 0x31, 0x62, 0xc4, 0x95, 0x37, 0x6e, 0xdc, 0xa5, 0x57, 0xae, 0x41, 0x82,
-    0x19, 0x32, 0x64, 0xc8, 0x8d, 0x07, 0x0e, 0x1c, 0x38, 0x70, 0xe0, 0xdd, 0xa7, 0x53, 0xa6, 0x51,
-    0xa2, 0x59, 0xb2, 0x79, 0xf2, 0xf9, 0xef, 0xc3, 0x9b, 0x2b, 0x56, 0xac, 0x45, 0x8a, 0x09, 0x12,
-    0x24, 0x48, 0x90, 0x3d, 0x7a, 0xf4, 0xf5, 0xf7, 0xf3, 0xfb, 0xeb, 0xcb, 0x8b, 0x0b, 0x16, 0x2c,
-    0x58, 0xb0, 0x7d, 0xfa, 0xe9, 0xcf, 0x83, 0x1b, 0x36, 0x6c, 0xd8, 0xad, 0x47, 0x8e, 0x01, 0x02,
+pub const PRIMITIVE_POLYS: &[u16] = &[
+    0x11B, 0x11D, 0x12B, 0x12D, 0x14D, 0x15F, 0x163, 0x165, 0x169, 0x171, 0x187, 0x1A9, 0x1C3,
+    0x1CF, 0x1E7, 0x1F5,
 ];
 
+#[inline]
+const fn is_primitive(poly: u16) -> bool {
+    let mut i = 0;
+
+    while i < PRIMITIVE_POLYS.len() {
+        if PRIMITIVE_POLYS[i] == poly {
+            return true;
+        }
+        i += 1;
+    }
+    false
+}
+
 #[derive(Debug, PartialEq, Clone)]
 #[cfg_attr(feature = "fuzzing", derive(Arbitrary))]
 #[cfg_attr(feature = "zeroize_memory", derive(Zeroize))]
 #[cfg_attr(feature = "zeroize_memory", zeroize(drop))]
-pub struct GF256(pub u8);
+pub struct GF256<const POLY: u16>(pub u8);
+
+pub struct Tables<const POLY: u16> {
+    pub log: [u8; 256],
+    pub exp: [u8; 512],
+}
+
+impl<const POLY: u16> Default for Tables<POLY> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+const fn is_primitive_element<const POLY: u16>(x: u8) -> bool {
+    const FACTORS: [u8; 3] = [3, 5, 17];
+
+    let mut i = 0;
+    while i < FACTORS.len() {
+        if gf256_pow(x, 255 / FACTORS[i], POLY) == 1 {
+            return false;
+        }
+        i += 1;
+    }
+    gf256_pow(x, 255, POLY) == 1
+}
+
+const fn find_generator<const POLY: u16>() -> u8 {
+    let mut i = 1u16;
+    while i <= 255 {
+        if is_primitive_element::<POLY>(i as u8) {
+            return i as u8;
+        }
+        i += 1;
+    }
+    panic!("No primitive element found");
+}
+
+impl<const POLY: u16> Tables<POLY> {
+    pub const fn new() -> Self {
+        assert!(is_primitive(POLY), "POLY must be primitive");
+
+        let mut log = [0u8; 256];
+        let mut exp = [0u8; 512];
+
+        let gen = find_generator::<POLY>();
+
+        let mut i = 0usize;
+        let mut x = 1u8;
+        while i < 255 {
+            exp[i] = x;
+            log[x as usize] = i as u8;
+            x = gf256_mul(x, gen, POLY);
+            i += 1;
+        }
+
+        let mut j = 255usize;
+        while j < 512 {
+            exp[j] = exp[j - 255];
+            j += 1;
+        }
+
+        Self { log, exp }
+    }
+}
+
+impl<const POLY: u16> GF256<POLY> {
+    const POLY_CHECK: () = assert!(is_primitive(POLY), "POLY must be primitive");
```

Comment:
does this also work to show it fails at comptime if you use a non-primitive root?